va/va_prot.h

```diff
@@ -0,0 +1,456 @@
+ * va_status = vaCreateConfig(va_dpy, VAProfileProtected,
+ *     VAEntrypointProtectedContent, attribs, 5, &config_id);
+ * CHECK_VASTATUS(va_status, "vaCreateConfig");
+ * \endcode
+ *
+ * Once the config is set up, we can create protected content session via
+ vaCreateProtectedSession().
+ * \code
+ * // Create a protected session
+ * VAProtectedSessionID crypto_session;
+ *
+ * va_status = vaCreateProtectedSession(va_dpy, config_id, &crypto_session);
+ * CHECK_VASTATUS(va_status, "vaCreateProtectedSession");
+ * \endcode
+ *
+ * \section api_pc_exec TEE communication via vaProtectedSessionExecute()
+ *
+ * Before start decryption/encryption, we may need to communicate with TEE to
+ * get encrypted decryption. We need to call vaProtectedSessionExecute() to
+ * get it. The following pseudo-code demonstrates getting session id via
+ * vaProtectedSessionExecute() as an example.
+ *
+ * \code
+ * uint32_t app_id = 0xFF;
+ * VABufferID buffer;
+ * VAProtectedSessionExecuteBuffer exec_buff = {0};
+ *
+ * exec_buff.function_id = GET_SESSION_ID
+ * exec_buff.input.data = nullptr;
+ * exec_buff.input.data_size = 0;
+ * exec_buff.output.data = &app_id;
+ * exec_buff.output.max_data_size = (uint32_t);
+ * va_status = vaCreateBuffer(
+ *                 va_dpy,
+ *                 crypto_session,
+ *                 (VABufferType) VAProtectedSessionExecuteBufferType,
+ *                 sizeof(exec_buff),
+ *                 1,
+ *                 &exec_buff,
+ *                 &buffer);
+ *
+ * va_status = vaProtectedSessionExecute(va_dpy, crypto_session, buffer);
+ *
+ * vaDestroyBuffer(va_dpy, buffer);
+ * \endcode
+ *
+ * \section api_pc_attach Attach/Detach protected content session to the VA
+ * context which want to enable/disable decryption/protection
+ *
+ * Protected content session is attached to VA decode/encode/vp context to
+ * enable decryption/protection. It could do attaching before decoding/encoding
+ * each frame and detach after each frame processed
+ *
+ * \code
+ * vaAttachProtectedSession(va_dpy, decode_ctx, crypto_session);
+ * foreach (iteration) {
+ *     vaBeginPicture(va_dpy, decode_ctx, surface);
+ *     ...
+ *     vaRenderPicture(va_dpy, decode_ctx, &buf_id1, 1);
+ *     vaRenderPicture(va_dpy, decode_ctx, &buf_id2, 1);
+ *     ...
+ *     vaEndPicture(va_dpy, decode_ctx);
+ * }
+ * vaDetachProtectedSession(va_dpy, decode_ctx);
+ * \endcode
+ *
+ * or it could be frame-by-frame attaching/detaching as following:
+ *
+ * \code
+ * foreach (iteration) {
+ *     if (encrypted)
+ *         vaAttachProtectedSession(va_dpy, decode_ctx, crypto_session);
+
+ *     vaBeginPicture(va_dpy, decode_ctx, surface);
+ *     ...
+ *     vaRenderPicture(va_dpy, decode_ctx, &buf_id1, 1);
+ *     vaRenderPicture(va_dpy, decode_ctx, &buf_id2, 1);
+ *     ...
+ *     vaEndPicture(va_dpy, decode_ctx);
+ *
+ *     if (encrypted)
+ *         vaDetachProtectedSession(va_dpy, decode_ctx);
+
+ *     // check encrypted variable for next frame
```

Comment:
That's end of the example, but I still don't get it why VAEncryptionParameterBufferType is needed. It is missing in the example entirely.

Reply:
Updated.